### PR TITLE
Us 2.4

### PR DESCRIPTION
**Pull request by AbdelrahmanWM**
# Inter-Campus Routing & Shuttle Integration

## Overview

This PR implements inter-campus routing with automatic campus switching, enhances shuttle integration, and adds a comprehensive 3AAA unit testing suite.

---

###  Routing & Navigation

* Improved SGW/Loyola auto-switching.
* Fixed stale route persistence by ensuring proper `Polyline` unmounting and state clearing.

###  UI/UX Improvements

* Implemented dynamic **Start/Destination confirmation labels**.
* Added real-time **shuttle arrival hints**.

###  State Management

* Added race condition protection for directions handling.
* Implemented request cancellation to prevent stale responses.

###  Testing

* Added **15+ unit tests** for:

  * `DirectionsContext`
  * `MapView`
  * `BuildingInformation`
* Covered edge cases including:

  * Stale responses
  * Rapid input changes
  * Route clearing behavior

---

## Verification

* **Unit Tests:** 46/49 passing (`npm test`).
* **Manual Testing:**

  * Confirmed auto-campus switching.
  * Verified route clearing logic.
  * Validated shuttle hint accuracy.

---

## Notes

This PR completes #83 #84 #85 , establishing a robust foundation for inter-campus directions and map interactions.

### CODE REVIEW

File: campus-guide/app/context/DirectionsContext.tsx [hunk 1]
```diff
@@ -41,35 +41,67 @@ export default function DirectionsProvider({
...
+    activeRequestRef.current++; // Invalidate any pending requests
     setStartBuilding(null);
     setDestinationBuilding(null);
     setTransportationMode("walk");
     setRoute(null);
+    setIsLoadingRoute(false);
   }, []);
 
   const swapLocations = useCallback(() => {
+    console.log('Swapping locations');
```
Comment (antholim) [inline campus-guide/app/context/DirectionsContext.tsx]: Here as well

File: campus-guide/app/components/BuildingInformation.tsx [hunk 2]
```diff
@@ -15,6 +15,7 @@ interface BuildingInformationProps {
   readonly building: Building;
   readonly onGetDirections: (building: Building) => void;
   readonly onClose: () => void;
+  readonly getDirectionsButtonText?: string;
```
Comment (antholim) [inline campus-guide/app/components/BuildingInformation.tsx]: Do we need to make this nullable?

File: campus-guide/app/components/DirectionsHeader.tsx [hunk 3]
```diff
@@ -28,6 +29,43 @@ export default function DirectionsHeader() {
...
     const [searchQuery, setSearchQuery] = useState('');
     const [showScheduleModal, setShowScheduleModal] = useState(false);
+    const [nextShuttleTime, setNextShuttleTime] = useState<string | null>(null);
+
+    // Calculate next shuttle when mode is shuttle and campuses are different
+    useEffect(() => {
+        if (transportationMode === 'shuttle' && startBuilding && destinationBuilding && startBuilding.campus !== destinationBuilding.campus) {
+            const now = new Date();
+            const day = now.getDay();
+            const isFriday = day === 5;
```
Comment (antholim) [inline campus-guide/app/components/DirectionsHeader.tsx]: To me, those seem like magic numbers, could we have a constant file dedicated for that?

File: campus-guide/app/context/DirectionsContext.tsx
Comment (antholim) [inline campus-guide/app/context/DirectionsContext.tsx]: For this file, we should consider replacing console.log and console.error by a more comprehensive custom or premade logger.

File: campus-guide/app/components/DirectionsHeader.tsx [hunk 4]
```diff
@@ -6,6 +6,7 @@ import { TransportationMode } from '../types/transportation';
 import { SGW_BUILDINGS, LOYOLA_BUILDINGS, Building } from './../../constants/buildings';
 import useUserLocation from '../hooks/useUserLocation';
 import ShuttleSchedule from './ShuttleSchedule';
+import { SHUTTLE_SCHEDULE } from './../../constants/shuttleSchedule';
```
[the remark this answers is not in the document]
Comment (AbdelrahmanWM) [reply campus-guide/app/components/DirectionsHeader.tsx]: good idea! added some aliases

[the remark this answers is not in the document]
Comment (AbdelrahmanWM) [reply campus-guide/app/components/DirectionsHeader.tsx]: can you approve if you don't have any further comments!
